### PR TITLE
Add dialogs from graph

### LINKED ISSUE #28: Communication between FrontEnd and BackEnd
In order to be story/composition agnostic and have fast development cycles we need to think of a communication language on top of socket io - e.g. it would be good to implement certain information retrieval tools like

* TextBox
* Selection Box
* GPS location

This can be activated/requested/displayed via socketIO - is this possible or is this hard @vin-ni ? 🍡 

The backend then plays the story/composition and the frontend is only there for data exchange. This has the advantage that the whole state of the story only needs to be managed from the backend and also to not leak story elements via JS to the user.

### PR DESCRIPTION
**Pull request by capital-G**
Closes #28 

image

results in a pop up

image

## ToDo

* [x] ~~Separate stream info and stream instruction~~ => #483 
* [x] Fix and write tests for backend
* [x] Fix styling
* [x] Write backend tests
* [x] Documentation

### CODE REVIEW

**Comment by capital-G** [comment]
Ready for review @vin-ni 

Style needs fixing and I don't know how to fix this.
Also I am unsure if I broke the Drifter template - maybe we can deprecate this with this? I think only the GPS is missing but this shouldn't be too hard to implement as an addition.

Docs are also a first draft but at least there exist something on how to use the script cells properly.

I also don't really like the current code-layout of `GraphPlayer` as it does not really do a separation of concerns, but due its heavy interconnection I found this the cleanest way to implement without relying too much on emits and mutating props.

**Comment by capital-G** [comment]
Now with GPS dialog as well by adding a "callback action" to checkboxes or buttons :)

image

Can be triggered as easy as

image

Ready for review @vin-ni :)

File: caster-back/story_graph/engine.py [hunk 1]
```diff
@@ -42,8 +47,14 @@ def __init__(
...
-        """Could be a @property but this can be difficult in async contexts
-        so we use explicit async via a getter method.
+        """
+        Returns the associated :class:`~stream.models.StreamVariable` within
+        this :class:`~stream.models.Stream` session.
+
+        .. todo::
+
+            Could be a @property but this can be difficult in async contexts
+            so we use explicit async via a getter metho
```
Comment (vin-ni) [inline caster-back/story_graph/engine.py:57]: typo: `metho` -> `method`

File: caster-back/stream/frontend_types.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
+"""
+These types are not reflecting any database content but are for triggering
+functionality in the frontend from within a :class`~story_graph.models.Graph`.
+
+These types can be used within a Python :class:`~story_graph.models.ScriptCell`
+as they are also available within the :class:`~stream.engine.Engine`.
+
+The stream subscription makes it possible to yield the
```
Comment (vin-ni) [inline caster-back/stream/frontend_types.py:8]: missing end of sentence?

File: caster-front/src/components/GraphDetailTemplates/DrifterTemplate.vue [hunk 3]
```diff
@@ -197,17 +212,17 @@ setTimeout(() => {
...
-              :stream-uuid="data.streamInfo.stream.uuid"
+              :stream-uuid="streamInfo.stream.uuid"
               @submitted="() => shiftDialogs()"
             />
           </div>
         </div>
 
         <!-- player -->
-        <div v-if="data.streamInfo.stream">
+        <div v-if="streamInfo.stream">
```
Comment (vin-ni) [inline caster-front/src/components/GraphDetailTemplates/DrifterTemplate.vue:222]: Thank you for cleaning up!

File: caster-front/src/components/GraphPlayer.vue [hunk 4]
```diff
@@ -1,27 +1,272 @@
...
+> = ref(undefined);
+
+const currentDialog = computed<DialogShow | undefined>(() => {
+  if (dialogsToShow.value.length < 1) {
+    return undefined;
+  }
+  return dialogsToShow.value[0];
 });
 
+// send a start variable when the client is connected to a stream
```
Comment (vin-ni) [inline caster-front/src/components/GraphPlayer.vue:70]: Great!

File: docs/editor.rst [hunk 5]
```diff
@@ -1,4 +1,112 @@
...
+        content=[
+            Text(text="Can we ask for your name?"),
+            Input(label="Name", key="name"),
+        ],
+        buttons=[Button.cancel(), Button.ok()],
+    )
+    # wait for user input
+    await wait_for_stream_variable("name")
+
+This will save the value that was inserted by the user into tho variable ``name``.
```
Comment (vin-ni) [inline docs/editor.rst:85]: This is so great. We can also have custom templates, opening dialogs with buttons like:
"change name" or so. Just a thought I wanted to share :)

File: docs/editor.rst [hunk 6]
```diff
@@ -1,4 +1,112 @@
...
+.. code-block:: supercollider
+
+    Ndef(\drone, {
+        SinOsc.ar(LFDNoise1.kr(0.1!2).exprange(100, 400)) * \amp.kr(0.2);
+    }).fadeTime_(10.0).play;
+    10.0.wait;
+
+.. important::
+
+    Do not execute ``Tdef.clear`` as this will remove the beacon of the server and
```
Comment (vin-ni) [inline docs/editor.rst:105]: Can we catch this ?

**Review by vin-ni — COMMENTED** [review]
Incredible!
Tiny changes otherwise fine to merge once i fixed the styling issues

**Comment by vin-ni** [comment]
Screenshot 2023-07-18 at 19 30 45

Added styling

**Comment by capital-G** [comment]
So what is the current state of this @vin-ni ? :D Sorry, somehow I missed all the comments, I now turned all notifications on again.

**Comment by vin-ni** [comment]
@capital-G it's ready to be merged :)

**Pull request merged**